homeassistant/components/elmax/__init__.py

```diff
@@ -0,0 +1,284 @@
+"""The elmax-cloud integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from logging import Logger
+from typing import Any, Mapping
+
+import async_timeout
+from elmax_api.exceptions import (
+    ElmaxApiError,
+    ElmaxBadLoginError,
+    ElmaxBadPinError,
+    ElmaxNetworkError,
+)
+from elmax_api.http import Elmax
+from elmax_api.model.endpoint import DeviceEndpoint
+from elmax_api.model.panel import PanelEntry, PanelStatus
+
+from homeassistant.components.elmax.const import (
+    CONF_ELMAX_PANEL_ID,
+    CONF_ELMAX_PANEL_PIN,
+    CONF_ELMAX_PASSWORD,
+    CONF_ELMAX_USERNAME,
+    DEFAULT_TIMEOUT,
+    DOMAIN,
+    ELMAX_PLATFORMS,
+    POLLING_SECONDS,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.exceptions import ConfigEntryAuthFailed, HomeAssistantError
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up elmax-cloud from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    # Retrieve username/password from integration configuration
+    username: str = entry.data[CONF_ELMAX_USERNAME]
+    password: str = entry.data[CONF_ELMAX_PASSWORD]
+    panel_id: str = entry.data[CONF_ELMAX_PANEL_ID]
+    panel_pin: str = entry.data[CONF_ELMAX_PANEL_PIN]
+
+    # Create the API client object and attempt a login, so that we immediately know
+    # if there is something wrong with user credentials
+    coordinator = ElmaxCoordinator(
+        hass=hass,
+        logger=_LOGGER,
+        username=username,
+        password=password,
+        panel_id=panel_id,
+        panel_pin=panel_pin,
+        name=f"Elmax Cloud {entry.entry_id}",
+        update_interval=timedelta(seconds=POLLING_SECONDS),
+    )
+
+    # Issue a first refresh, so that we trigger a re-auth flow if necessary
+    await coordinator.async_config_entry_first_refresh()
+
+    # Store a global reference to the coordinator for later use
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    # Perform platform initialization.
+    # Add the cover feature only if supported by the current panel.
+    plats = ELMAX_PLATFORMS.copy()
+    if coordinator.panel_status is not None and coordinator.panel_status.cover_feature:
+        plats.append("cover")
+    hass.config_entries.async_setup_platforms(entry, plats)
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, ELMAX_PLATFORMS)
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class ElmaxCoordinator(DataUpdateCoordinator):
```

Comment:
I think it is clearer if you move this class to a separate file (eg. `common.py` or `elmax_coordinator.py`)

Reply:
Agreed!